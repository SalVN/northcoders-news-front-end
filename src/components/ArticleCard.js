import React from 'react';
import PropTypes from 'prop-types';
import { BrowserRouter as Router, Link } from 'react-router-dom';

const ArticleCard = function (props) {
  return (
    <div className='box'>
      <article className='media'>
        <div className='media-left'>
          {props.number + 1}
        </div>
        <div className='media-left'>
          <button className='button' onClick={props.voteHandler.bind(this, 'up', props.id)}><i className="fa fa-arrow-up" aria-hidden="true"></i></button>
          <div>{props.votes}</div>
          <button className='button' onClick={props.voteHandler.bind(this, 'down', props.id)}><i className="fa fa-arrow-down" aria-hidden="true"></i></button>
        </div>
        <div className='media-content'>
          <div className='content'>
            <Link to={`/articles/${props.id}`}>
              <h3 className='title is-3'>{props.title}</h3>
            </Link>
            <p>
              <span>added by: {props.author}</span>
              <span>tags: {props.tags}</span>
              <span>comments: {props.comment_count}</span>
            </p>
          </div>
        </div>
      </article>
    </div>
  );
};

ArticleCard.propTypes = {
  title: PropTypes.string.isRequired,
  votes: PropTypes.number.isRequired,
  number: PropTypes.number.isRequired,
  author: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  tags: PropTypes.string.isRequired,
  comment_count: PropTypes.number.isRequired,
  voteHandler: PropTypes.func.isRequired
};


export default ArticleCard;
