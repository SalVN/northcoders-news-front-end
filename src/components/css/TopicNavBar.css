.navbar-css {
    border-bottom: 2px solid #373737;
}

.navbar-text-css {
    color: black !important;
    font-weight: strong !important;
}

.logo {
    height: 30px;
    width: auto;
}

.avatar {
    border-radius: 50px;
    height: 25px;
    margin: 5px;
}

.user {
    background-color: #373737;
;
    border-radius: 5px;
    padding: 5px;
    color: white !important;
    height: 40px;
    display: flex;
    align-items: center;
}

.navbar-user-text {
    margin: 5px;
}
