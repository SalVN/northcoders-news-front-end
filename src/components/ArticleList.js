import React from 'react';
import ArticleCard from './ArticleCard';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import * as actions from '../actions/actions';

class ArticleList extends React.Component {
  componentDidMount() {
    if (this.props.users && this.props.users.length < 1) {
      this.props.fetchUsers();
    }
  }
  render() {
    const users = this.props.users;
    const voteHandler = this.props.voteArticle;
    return (
      <div id='ArticleList' className='content'>
        {this.props.articles.length > 0 &&
          this.props.articles.map((article, i) => {
            const username = article.created_by;
            let index;
            users &&
              users.forEach((user, i) => {
                if (user.username === username) index = i;
              });
            return (
              <ArticleCard
                key={article._id}
                title={article.title}
                number={i}
                votes={article.votes}
                author={article.created_by}
                id={article._id}
                tags={article.belongs_to}
                comment_count={article.comment_count}
                voteHandler={voteHandler}
                userData={users[index]}
              />
            );
          })
        }
      </div>
    );
  }
}

function mapDispatchToProps(dispatch) {
  return {
    fetchUsers: () => {
      dispatch(actions.fetchUsers());
    }
  };
}

function mapStateToProps(state) {
  return {
    loading: state.users.loading,
    users: state.users.users
  };
}

ArticleList.propTypes = {
  articles: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  fetchUsers: PropTypes.func.isRequired,
  voteArticle: PropTypes.func.isRequired,
  users: PropTypes.array.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(ArticleList);