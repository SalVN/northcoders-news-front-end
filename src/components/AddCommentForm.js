import React from 'react';
import PropTypes from 'prop-types';

import './css/AddCommentForm.css';

class AddCommentForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            text: ''
        };
        this.handleChange = this.handleChange.bind(this);
    }
    render() {
        return (
            <div>
                {this.props.showForm
                    ?
                    <div className='box box-style box-style-comment box-style-comment'>
                        <article className='media'>
                            <div className='media-left'>
                                <img className='comment-form-image' src={this.props.user.avatar_url} />
                            </div>
                            <div className='media-content'>
                                <div className='content'>
                                    <form onSubmit={this.props.handleSubmit}>
                                        <label className='comment-form-heading' htmlFor="comment-input">Join the conversation:</label>
                                        <br />
                                        <div className='comment-form' onBlur={this.props.toggleForm.bind(this, this.state.text)}>
                                            <textarea className='add-comment-textarea' onChange={this.handleChange} id="comment-input" defaultValue={this.state.text} placeholder='Join the conversation' />
                                        </div>
                                        <div className='button-div'>
                                            <button className='button add-comment-submit' type='submit'>Submit</button>
                                        </div>
                                    </form>
                                </div>
                            </div>
                        </article>
                    </div>
                    : <div className='box box-style box-style-comment box-style-comment-small'>
                        <article className='media'>
                            <div className='media-left'>
                                <img className='comment-form-image' src={this.props.user.avatar_url} />
                            </div>
                            <div className='media-content'>
                                <div className='content'>
                                    <div onClick={this.props.toggleForm.bind(this, null)} className='comment-form-small'>
                                        <p>Join the conversation</p>
                                    </div>
                                </div>
                            </div>
                        </article>
                    </div>
                }
            </div>
        );
    }
    handleChange(e) {
        this.setState({
            text: e.target.value
        });
    }
}

AddCommentForm.propTypes = {
    id: PropTypes.string.isRequired,
    addComment: PropTypes.func.isRequired,
    toggleForm: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired,
    showForm: PropTypes.bool.isRequired,
    handleSubmit: PropTypes.func.isRequired
};

export default AddCommentForm;