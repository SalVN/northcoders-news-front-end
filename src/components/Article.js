import React from 'react';
import PropTypes from 'prop-types';
import { BrowserRouter as Router, Link } from 'react-router-dom';

import './css/Article.css';

const Article = function (props) {
    return (
        <div className='box'>
            <div className='columns'>
                <div className='column is-1'>
                    <button className='button is-black is-inverted is-small'><i className="fa fa-arrow-up" aria-hidden="true"></i></button>
                    <p className='comment-card-votes'><strong>{props.article.votes}</strong></p>
                    <button className='button is-black is-inverted is-small'><i className="fa fa-arrow-down" aria-hidden="true"></i></button>
                </div>
                <div className='column'>
                    <div className='title is-3 article-title'>{props.article.title}</div>
                    <p>
                        <span>
                            <span className='article-subheading-key'>added by:</span>
                            <Link to='#'>
                                <span>
                                    {props.user &&
                                        <img className='article-card-avatar' src={props.user.avatar_url} alt='user avatar' />
                                    }
                                </span>
                                <span className='article-subheading-value'>
                                    <strong>{props.article.created_by}</strong>
                                </span>
                            </Link>
                        </span>
                        <span>
                            <span className='article-card-subheading-key'>tags:</span>
                            <Link to={`/topics/${props.article.belongs_to}/articles`}>
                                <span className='article-card-subheading-value'>
                                    <strong>{props.article.belongs_to}</strong>
                                </span>
                            </Link>
                        </span>
                        <span>
                            <span className='article-card-subheading-key'>comments:</span>
                            {/*<Link to='#'>*/}
                            <span className='article-card-subheading-value-notlink'>
                                <strong>{props.article.commentCount}</strong>
                            </span>
                            {/*</Link>*/}
                        </span>
                    </p>
                    <div>{props.article.body}</div>
                </div>
            </div>
        </div>
    );
};

Article.propTypes = {
    article: PropTypes.object.isRequired,
    user: PropTypes.object.isRequired
};

export default Article;