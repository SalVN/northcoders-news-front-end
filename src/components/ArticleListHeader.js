import React from 'react';
import PropTypes from 'prop-types';

import './css/MainArticleList.css';

const ArticleListHeader = function (props) {
    return (
        <div className='content'>
            <div className='columns'>
                <div className='column is-3'>
                    <h2 className='page-title'><strong>{props.title}</strong></h2>
                </div>
                <div className='column is-1 is-offset-5 label'>
                    <label htmlFor="Sort by">Sort by:</label>
                </div>
                <div className='column is-3'>
                    <div className='select is-primary select-menu'>
                        <select name="Sort by" onChange={props.handleClickSelect}>
                            <option value="votes">Popularity</option>
                            <option value="comments">Number of Comments</option>
                            <option value="random">Random</option>
                        </select>
                    </div>
                </div>
            </div>
        </div>
    );
};

ArticleListHeader.propTypes = {
    title: PropTypes.string.isRequired,
    handleClickSelect: PropTypes.func.isRequired
};

export default ArticleListHeader;