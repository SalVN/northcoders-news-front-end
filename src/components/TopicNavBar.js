import React, { Component } from 'react';
import { BrowserRouter as Router, NavLink, Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import * as actions from '../actions/actions';
import './css/TopicNavBar.css';
import { USERNAME } from '../../config';

class TopicNavBar extends Component {
    constructor(props) {
        super(props);
        this.state = {
            activeBurger: false
        };
        this.toggleBurger = this.toggleBurger.bind(this);
    }
    componentDidMount() {
        this.props.fetchTopics();
        this.props.fetchUser(USERNAME);
    }
    render() {
        let activeBurger = (this.state.activeBurger) ? 'is-active' : '';
        return (
            <nav className='navbar navbar-css'>
                <div className='navbar-brand'>
                    <span className='navbar-item' >
                        <Link to='/'>
                            <img className='logo' src='https://avatars3.githubusercontent.com/u/6791502?v=3&s=200' alt='Northcoders logo' />
                        </Link>
                    </span>
                    <div onClick={this.toggleBurger} className={`navbar-burger ${activeBurger}`}>
                        <div className="navbar-burger">
                            <span></span>
                            <span></span>
                            <span></span>
                        </div>
                    </div>
                </div>
                <div className={`navbar-menu ${activeBurger}`}>
                    <div className='navbar-start'>
                        {
                            this.props.topics &&
                            this.props.topics.map((topic) => {
                                return (
                                    <span key={topic._id} className='navbar-item is-hoverable'>
                                        <NavLink className='navbar-text-css' to={`/topics/${topic.slug}/articles`}>
                                            <span>{`< ${topic.title.toUpperCase()} />`}</span>
                                        </NavLink>
                                    </span>
                                );
                            })}
                    </div>
                    <div className='navbar-end'>
                        <span className='navbar-item'>
                            <Link className='user' to='#'>
                                <span><img className='avatar' src={`${this.props.user.avatar_url}`} /></span>
                                <span className='navbar-user-text'><strong>{this.props.user.username}</strong></span>
                            </Link>
                        </span>
                    </div>
                </div>
            </nav >
        );
    }
    toggleBurger() {
        this.setState({
            activeBurger: !this.state.activeBurger
        });
    }
}

function mapDispatchToProps(dispatch) {
    return {
        fetchTopics: () => {
            dispatch(actions.fetchTopics());
        },
        fetchUser: (username) => {
            dispatch(actions.fetchOneUser(username));
        }
    };
}

function mapStateToProps(state) {
    return {
        topics: state.topics.topics,
        topicsLoading: state.topics.loading,
        userLoading: state.oneUser.loading,
        user: state.oneUser.user
    };
}

TopicNavBar.propTypes = {
    topics: PropTypes.array.isRequired,
    loading: PropTypes.bool.isRequired,
    fetchTopics: PropTypes.func.isRequired,
    fetchUser: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(TopicNavBar);