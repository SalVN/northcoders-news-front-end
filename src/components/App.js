import React, {Component} from 'react';
import PropTypes from 'prop-types';

import TopicNavBar from './TopicNavBar';
import MainBanner from './MainBanner';
import UserCard from './UserCard';
import UserListCard from './UserListCard';

import './css/App.css';

export class App extends Component {
  render() {
    let pageStyling = window.innerWidth < 800 ? 'main-page-mobile' : 'main-page';
    return (
      <div className='page'>
        <TopicNavBar />
        <MainBanner />
        <div className={pageStyling}>
          <div className={'columns is-gapless'}>
            <div className='column is-three-quarters'>
              {this.props.children}
            </div>
            <div className='column is-one-quarter'>
              <div className='is-hidden-mobile'>
                <UserCard />
              </div>
              <UserListCard />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

App.propTypes = {
  children: PropTypes.element.isRequired
};

export default App;
