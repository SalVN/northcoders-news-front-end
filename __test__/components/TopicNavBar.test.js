import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';
import sinon from 'sinon';

const mockStore = configureStore();
const initialState = {};

import { TopicNavBar } from '../../src/components/TopicNavBar';

describe('TopicNavBar', () => {
    const topics = [{
        _id: '59b04728371c52f1739aba0b',
        title: 'Football',
        slug: 'football',
        __v: 0
    },
    {
        _id: '59b04728371c52f1739aba0c',
        title: 'Cooking',
        slug: 'cooking',
        __v: 0
    },
    {
        _id: '59b04728371c52f1739aba0d',
        title: 'Cats',
        slug: 'cats',
        __v: 0
    }];
    const user = {
        _id: '59b1b18b327cce1fb043bdb1',
        username: 'northcoder',
        name: 'Awesome Northcoder',
        avatar_url: 'https://avatars3.githubusercontent.com/u/6791502?v=3&s=200',
        __v: 0
    };
    it('is a function', () => {
        expect(typeof TopicNavBar).toEqual('function');
    });

    it('renders', () => {
        const store = mockStore(initialState);
        const wrapper = shallow(<TopicNavBar
            topics={topics}
            topicsLoading={false}
            userLoading={false}
            fetchTopics={x => x}
            fetchUser={x => x}
            user={user}
            store={store}
        />);
        expect(wrapper.children().length).toEqual(2);
        expect(wrapper.find('nav').length).toEqual(1);
    });

    it('renders correctly', () => {
        const store = mockStore(initialState);
        const tree = renderer.create(
            <Provider store={store}>
                <MemoryRouter>
                    <TopicNavBar
                        topics={topics}
                        topicsLoading={false}
                        userLoading={false}
                        fetchTopics={x => x}
                        fetchUser={x => x}
                        user={user}
                        store={store}
                    />
                </MemoryRouter>
            </Provider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('has an initial state', () => {
        const store = mockStore(initialState);
        const wrapper = shallow(<TopicNavBar
            topics={topics}
            topicsLoading={false}
            userLoading={false}
            fetchTopics={x => x}
            fetchUser={x => x}
            user={user}
            store={store}
        />);
        expect(wrapper.state()).toEqual({ activeBurger: false });
    });

    it('fetches the topics and user on componentDidMount', () => {
        const spyTopics = sinon.stub();
        const spyUser = sinon.stub();
        const store = mockStore(initialState);
        mount(
            <MemoryRouter>
                <TopicNavBar
                    topics={topics}
                    topicsLoading={false}
                    userLoading={false}
                    fetchTopics={spyTopics}
                    fetchUser={spyUser}
                    user={user}
                    store={store}
                />
            </MemoryRouter>);
        expect(spyTopics.callCount).toBe(1);
        expect(spyUser.callCount).toBe(1);
    });

    it('changes the burger to "is-active" if this.state.activeBurger', () => {
        const enzymeWrapper = shallow(<TopicNavBar
            topics={topics}
            topicsLoading={false}
            userLoading={false}
            fetchTopics={x => x}
            fetchUser={x => x}
            user={user}
            store={x => x}
        />);
        expect(enzymeWrapper.find('.navbar-menu').node.props.className).toBe('navbar-menu ');
        enzymeWrapper.setState({ activeBurger: true });
        expect(enzymeWrapper.find('.navbar-menu').node.props.className).toBe('navbar-menu is-active');
    });

    it('contains two Links', () => {
        const enzymeWrapper = shallow(<TopicNavBar
            topics={topics}
            topicsLoading={false}
            userLoading={false}
            fetchTopics={x => x}
            fetchUser={x => x}
            user={user}
            store={x => x}
        />);
        expect(enzymeWrapper.find('Link').length).toBe(2);
    });

    it('contains three NavLinks', () => {
        const enzymeWrapper = shallow(<TopicNavBar
            topics={topics}
            topicsLoading={false}
            userLoading={false}
            fetchTopics={x => x}
            fetchUser={x => x}
            user={user}
            store={x => x}
        />);
        expect(enzymeWrapper.find('NavLink').length).toBe(3);
    });

    it('triggers the toggleBurger function when the navbar-burger is clicked', () => {
        const enzymeWrapper = shallow(<TopicNavBar
            topics={topics}
            topicsLoading={false}
            userLoading={false}
            fetchTopics={x => x}
            fetchUser={x => x}
            user={user}
            store={x => x}
        />);
        expect(enzymeWrapper.state('activeBurger')).toBe(false);
        enzymeWrapper.find('.burger').simulate('click');
        expect(enzymeWrapper.state('activeBurger')).toBe(true);
    });

    it('shows a loading icon if this.props.topicsLoading', () => {
        const enzymeWrapperA = shallow(<TopicNavBar
            topics={topics}
            topicsLoading={false}
            userLoading={false}
            fetchTopics={x => x}
            fetchUser={x => x}
            user={user}
        />);
        expect(enzymeWrapperA.find('.fa-spin').length).toBe(0);

        const enzymeWrapperB = shallow(<TopicNavBar
            topics={topics}
            topicsLoading={true}
            userLoading={false}
            fetchTopics={x => x}
            fetchUser={x => x}
            user={user}
            store={x => x}
        />);
        expect(enzymeWrapperB.find('.fa-spin').length).toBe(1);
    });

    it('shows a loading icon if this.props.userLoading', () => {
        const store = mockStore(initialState);
        const enzymeWrapperA = shallow(<TopicNavBar
            topics={topics}
            topicsLoading={false}
            userLoading={false}
            fetchTopics={x => x}
            fetchUser={x => x}
            user={user}
            store={x => x}
        />);
        expect(enzymeWrapperA.find('.fa-spin').length).toBe(0);

        const enzymeWrapperB = shallow(<TopicNavBar
            topics={topics}
            topicsLoading={false}
            userLoading={true}
            fetchTopics={x => x}
            fetchUser={x => x}
            user={user}
            store={x => x}
        />);
        expect(enzymeWrapperB.find('.fa-spin').length).toBe(1);
    });

    it('only renders the topics nav-links if topics is defined', () => {
        const enzymeWrapperA = shallow(<TopicNavBar
            topics={[]}
            topicsLoading={false}
            userLoading={false}
            fetchTopics={x => x}
            fetchUser={x => x}
            user={user}
        />);
        expect(enzymeWrapperA.find('NavLink').length).toBe(0);
    });

    it('renders the same number of NavLinks as there are topics in the array', () => {
        const enzymeWrapperA = shallow(<TopicNavBar
            topics={topics}
            topicsLoading={false}
            userLoading={false}
            fetchTopics={x => x}
            fetchUser={x => x}
            user={user}
        />);
        expect(enzymeWrapperA.find('NavLink').length).toBe(3);
    });
});