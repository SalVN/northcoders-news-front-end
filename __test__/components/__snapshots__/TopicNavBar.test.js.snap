// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopicNavBar renders correctly 1`] = `
<nav
  className="navbar navbar-css"
>
  <div
    className="navbar-brand"
  >
    <span
      className="navbar-item"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <img
          alt="Northcoders logo"
          className="logo"
          src="https://avatars3.githubusercontent.com/u/6791502?v=3&s=200"
        />
      </a>
    </span>
    <div
      className="navbar-burger "
      onClick={[Function]}
    >
      <div
        className="navbar-burger"
      >
        <span />
        <span />
        <span />
      </div>
    </div>
  </div>
  <div
    className="navbar-menu "
  >
    <div
      className="navbar-start"
    />
    <div
      className="navbar-end"
    >
      <span
        className="navbar-item"
      />
    </div>
  </div>
</nav>
`;
